### PR TITLE
Autogenererating enum-string mappings from grammar

### PR DESCRIPTION
**Pull request by atgoo**
Autogenerating the following code:
- extension enum
- extension-to-string
- string-to-extension
- capability-to-string

### CODE REVIEW

**Comment by atgoo** [comment]
I want to redo the cmake part. Will reupload.

File: test/enum_string_mapping_test.cpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+TEST_P(ExtensionTest, TestExtensionToString) {
+  const std::pair<Extension, std::string> param = GetParam();
+  const Extension extension = param.first;
+  const std::string extension_str = param.second;
+  const std::string result_str = libspirv::ExtensionToString(extension);
+  EXPECT_EQ(extension_str, result_str);
+}
+
+TEST_P(CapabilityTest, TestCapabilityToString) {
+  const std::pair<SpvCapability, std::string> param = GetParam();
```
Comment (dneto0) [inline test/enum_string_mapping_test.cpp]: This is ok, but could have been const ref.

File: test/enum_string_mapping_test.cpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+      {SpvCapabilityPerViewAttributesNV,
+       "PerViewAttributesNV"}
+    })));
+
+TEST_P(ExtensionTest, TestExtensionFromString) {
+  const std::pair<Extension, std::string> param = GetParam();
+  const Extension extension = param.first;
+  const std::string extension_str = param.second;
+  Extension result_extension;
+  ASSERT_TRUE(libspirv::GetExtensionFromString(extension_str, &result_extension));
```
Comment (dneto0) [inline test/enum_string_mapping_test.cpp]: Also need failure cases for GetExtensionFromString

File: test/enum_string_mapping_test.cpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+using ::libspirv::Extension;
+
+using ::testing::Values;
+using ::testing::ValuesIn;
+
+using ExtensionTest = ::testing::TestWithParam<std::pair<Extension, std::string>>;
+using CapabilityTest = ::testing::TestWithParam<std::pair<SpvCapability, std::string>>;
+
+INSTANTIATE_TEST_CASE_P(AllExtensions, ExtensionTest,
```
Comment (dneto0) [inline test/enum_string_mapping_test.cpp]: Typically we put the test class before the list of parameterized instances.

File: utils/generate_grammar_tables.py [hunk 4]
```diff
@@ -348,6 +348,91 @@ def generate_operand_kind_table(enums):
...
+    function += '  };\n\n  return "";\n}'
+    return function
+
+
+def generate_string_to_extension_mapping(operands):
+    """Returns mapping function from strings to corresponding extensions."""
+    extensions = get_extension_list(operands)
+    function = 'bool GetExtensionFromString(' \
+        'const std::string& str, Extension* extension) {\n '
+    template = ' if (str == "{extension}")\n' \
```
Comment (dneto0) [inline utils/generate_grammar_tables.py]: This bakes in the linear search.  If the number of extensions grows large then we might consider doing a regular STL lookup into an unordered map.  Suggest adding a TODO here.

At least this is no worse than the existing algorithm.

Comment (atgoo) [inline utils/generate_grammar_tables.py]: Could also be an array of strings, accessible by index.

Comment (atgoo) [inline utils/generate_grammar_tables.py]: The last comment was for extension->string.

Comment (dneto0) [inline utils/generate_grammar_tables.py]: Yes.  An optimizing compiler can turn a switch on a dense enum into a table lookup or binary search.  At least it used to be that way.  :-)

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as c804c125c

**Pull request closed without merging**